Add blog post Automatic Windows Installation With Tekton Pipelines (#727)
kubevirt/kubevirt.github.io: _posts/2021-04-19-Automatic-Windows-Installation-With-Tekton-Pipelines.md

---
layout: post

> author: Filip Křepinský

Reviewer: the spelling checks are falling due to the name
can you add it on the upstream
https://github.com/kubevirt/project-infra/blob/master/images/yaspeller/.yaspeller.json

Author: It might pass since I am already added to https://github.com/kubevirt/kubevirt.github.io/blob/source/.yaspeller.json

Reviewer: it's going to change to the upstream, nvm i need to create some changes there i will add you
 mark as resolved

Author: ok, thanks

description: This blog shows how KubeVirt Tekton Tasks can be utilized to automatically install and setup Windows VMs from scratch
navbar_active: Blogs
category: news
tags:
  [
    "kubevirt",
    "Kubernetes",
    "virtual machine",
    "VM",

>     "Tekton Pipelines"

Reviewer: missing comma 
```markdown
 "Tekton Pipelines",
```

Author: thanks

    "KubeVirt Tekton Tasks",
    "Windows"
  ]
comments: true
title: Automatic Windows Installation With Tekton Pipelines
pub-date: April 19
pub-year: 2021
---

# Introduction

This blog shows how we can easily automate a process of installing Windows VMs on KubeVirt with [Tekton Pipelines](https://github.com/tektoncd/pipeline).

Tekton Pipelines can be used to create a single `Pipeline` that encapsulates the installation process which can be run and replicated with `PipelineRuns`. 
The pipeline will be built with [KubeVirt Tekton Tasks](https://github.com/kubevirt/kubevirt-tekton-tasks) which include all the necessary tasks for this example.

## Pipeline Description

The pipeline will prepare an empty PVC (Persistent Volume Claim) and download a Windows source ISO into another PVC. Both of them will be initialized with CDI (Containerized Data Importer).

> It will then spin up a VM and use Windows Answer Files to automatically install the system.

Reviewer 2: I (think) you later refer to the term "installation VM", better define it here. Also, I find "system" as too vague.
```suggestion
It will then spin up an installation VM and use Windows Answer Files to automatically install the VM.
```

Author: Thanks, fixed

> Then the pipeline will wait for the installation to complete and will delete the installation VM while keeping the artifact PVC with the installed system.

Reviewer 2: ```suggestion
Then the pipeline will wait for the installation to complete and will delete the installation VM while keeping the artifact PVC with the installed operating system.
You can use the artifact PVC for ???
```

Author: fixed and added a new sentence: `You can later use the artifact PVC as a base image and copy it for new VMs`

    - taskName: create-vm-from-manifest
      serviceAccountName: create-vm-from-manifest-task
    - taskName: wait-for-vmi-status
      serviceAccountName: wait-for-vmi-status-task
    - taskName: cleanup-vm
      serviceAccountName: cleanup-vm-task
```

## Inspecting the output

Firstly, you can inspect the progress of the windows-10-source and windows-10-base import:

```bash
kubectl get dvs | grep windows-10-

> windows-10-base-8zxwr     Succeeded          100.0%                21s
> windows-10-source-jdv64   ImportInProgress   1.01%                 20s

```

To inspect the status of the pipeline run:

```bash
kubectl get pipelinerun -l "tekton.dev/pipeline=windows-installer"

> NAME                          SUCCEEDED   REASON                          STARTTIME   COMPLETIONTIME
> windows-installer-run-n2mjf   Unknown     Running                         118s
```

To check the status of each task and its pods:

```bash
kubectl get pipelinerun -o yaml -l "tekton.dev/pipeline=windows-installer"
kubectl get pods -l "tekton.dev/pipeline=windows-installer"
```

Once the pipeline run completes you should be left with `windows-10-base-xxxxx` PVC (backed by DataVolume).

> You can then create a VM with this DataVolume (`xxxxx` needs to be replaced in the following yaml) to test it.

Reviewer 2: I think you should suggest to first clone the PVC. If you don't intend to create multiple VMs from the same PVC, you could have just kept using the installation VM.

Author: good idea, changed the description and added a dataVolumeTemplate to the VM

              storage: 20Gi
        source:
          pvc:
            name: PVC_NAME
            namespace: PVC_NAMESPACE
  running: false
  template:
    metadata:
      labels:
        kubevirt.io/domain: windows-10-vm
    spec:
      domain:
        cpu:
          sockets: 2
          cores: 1
          threads: 1
        resources:
          requests:
            memory: 2Gi
        devices:
          disks:
            - name: rootdisk
              bootOrder: 1
              disk:
                bus: virtio
            - name: virtiocontainerdisk
              cdrom:
                bus: sata
          interfaces:
            - bridge: {}
              name: default
          inputs:
            - type: tablet
              bus: usb
              name: tablet
      networks:
        - name: default
          pod: {}
      volumes:
        - name: rootdisk
          dataVolume:
            name: windows-10-vm-root
        - name: virtiocontainerdisk
          containerDisk:
            image: kubevirt/virtio-container-disk
```

You can start the VM and login with `Administrator`:`changepassword` credentials. Then you should be welcomed to your fresh VM.

> ![Started Windows VM](/assets/2021-04-19-Automatic-Windows-Installation-With-Tekton-Pipelines/win-started.png "Started Windows VM")

Author: done and I doubled the thumbnail size to 200px - not sure what size is preferred?

# Resources

> - [YAML files used in this example](https://github.com/kubevirt/kubevirt-tekton-tasks/tree/main/examples/pipelines/windows-installer)

Reviewer 3: Hi! This appears to be a 404 not found.

Author: merged https://github.com/kubevirt/kubevirt-tekton-tasks/pull/80, so it should work now
